lib/logstash/codecs/cef.rb

```diff
@@ -1,89 +1,86 @@
-      data.sub! /^CEF:/, ''
-    end #if @syslog
-    # Now, break out the rest of the headers
-    event['cef_version'], event['cef_vendor'], event['cef_product'], event['cef_device_version'], event['cef_sigid'], event['cef_name'], event['cef_severity'], event['message'] =  data.scan /(?:[^\|\\]|\\.)+/
-    # Now, try to break out the Extension Dictionary
-    message=event['message']
-    if message.to_s.strip.length != 0
-      message = message.split(/ ([\w\.]+)=/)
-
-      key, value = message.shift.split('=',2)
-      @logger.debug(message)
-      kv = Hash[*message]
-      @logger.debug(kv)
-      addKey(kv,key,value)
-      event.to_hash.merge!(Hash[kv.map{ |k,v| ["cef_ext_"+k,v] }])
-    end #
-    yield event
-  end
-
-  public
-  def encode(data)
-    # "CEF:0|Elasticsearch|Logstash|1.0|Signature|Name|Sev|"
-
-    # TODO: Need to check that fields are set!
-
-    # Signature, Name, and Sev should be set in the config, with ref to fields
-    # Should also probably set the fields sent
-    header = ["CEF:0", "Elasticsearch", "Logstash", "1.0", @signature, @name, @sev].join("|")
-    values = @fields.map {|name| get_value(name, data)}.join(" ")
-    # values = values.map {|k,v| "#{k}=#{v}"}.join(" ")
-    @on_event.call(header + " " + values + "\n")
-  end
-
-  private
-  def addKey(kv_keys, key, value)
-    if kv_keys.has_key?(key)
-      if kv_keys[key].is_a? Array
-        kv_keys[key].push(value)
-      else
-        kv_keys[key] = [kv_keys[key], value]
-      end
-    else
-      kv_keys[key] = value
+    config_name "cef"
+    config :signature, :validate => :string, :default => "Logstash"
+    config :name, :validate => :string, :default => "Logstash"
+    config :sev, :validate => :number, :default => 6
+
+    config :fields, :validate => :array
+
+    public
+    def initialize(params={})
+        super(params)
     end
-  end # addKey
-
-  private
-  def get_value(name, event)
-    val = event[name]
-    case val
-    when Hash
-      return name + "=" + val.to_json
-    else
-      return name + "=" + val
+
+    public
+    def decode(data)
+        # Strip any quotations at the start and end, flex connectors seem to send this
+        if data[0] == "\""
+            data = data[0..-2]
+            data.slice!(0)
+        end
+        event = LogStash::Event.new()
+
+        # Split by the pipes
+        event['cef_version'], event['cef_vendor'], event['cef_product'], event['cef_device_version'], event['cef_sigid'], event['cef_name'], event['cef_severity'], message = data.split /(?<!\\)[\|]/
+
+        # Try and parse out the syslog header if there is one
+        if event['cef_version'].include? ' '
+            event['syslog'], event['cef_version'] = event['cef_version'].split(' ')
+        end
+
+        # Get rid of the CEF bit in the version
+        event['cef_version'].sub! /^CEF:/, ''
+
+        # Strip any whitespace from the message
+        message = message.to_s.strip
```

Comment:
I forgot to mention, I didn't implement this as if there was no "message" passed back in the CEF event (therefore no extension dictionary) then message would be nil which would result an error with the .strip, .to_s changes the nil to "" which then ensures the code after works out.

Could probably switch this out for a if statement asserting on nil I suppose, however I'm not sure what would be more performant on messages which do contain extra data (almost all), a `.to_s` or an `if not nil?`